### PR TITLE
Fix certain abilities not being overwritten by most forme changes

### PR DESCRIPTION
**Pull request by MacChaeger**
MnM test pending me figuring out how to do that.

### CODE REVIEW

**Comment by MacChaeger** [comment]
What.

**Comment by MacChaeger** [comment]
OK it seems the current test API doesn't support running tests in a specific format, only in custom formats with a mod. This doesn't work for MnM because the mixandmega mod relies on all Pokemon having a property that only gets set using the gen7mixandmega's onBegin function. Therefore, I'll hold off on the MnM test in this PR unless test API gets updated before this gets pushed.

**Comment by Zarel** [comment]
You shouldn't be writing MnM tests in the first place. It's unnecessary to write tests for formats that will stop working in a month.

**Comment by MacChaeger** [comment]
MnM is a permaladder...

**Comment by Zarel** [comment]
Well, then, uh, I didn't write the test framework. Feel free to update it.

**Comment by MacChaeger** [comment]
Same PR, or a new one?

**Comment by Zarel** [comment]
Up to you.

**Comment by MacChaeger** [comment]
Turns out I was just being an idiot, and it does in fact support tests by format.

File: test/simulator/misc/megaevolution.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+let battle;
+
+describe('Mega Evolution', function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it('should overwrite normally immutable abilities', function () {
+		battle = common.createBattle();
+		const p1 = battle.join('p1', 'Guest 1', 1, [{species: "Metagross", ability: 'zenmode', item: 'metagrossite', moves: ['metalclaw']}]);
```
Comment (Marty-D) [inline test/simulator/misc/megaevolution.js]: Zen Mode??? You had 9 options and you went off script!

File: test/simulator/misc/mixandmega.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+describe('Mix and Mega', function () {
+	beforeEach(function () {
+		battle = common.mod('gen7mixandmega').createBattle();
+	});
+	afterEach(() => battle.destroy());
+
+	it('should overwrite forme-change abilities on Mega Evolution or Primal Reversion', function () {
+		const p1 = battle.join('p1', 'Guest 1', 1, [{species: "Mimikyu", ability: 'disguise', item: 'metagrossite', moves: ['shadowclaw']}]);
+		const p2 = battle.join('p2', 'Guest 2', 1, [{species: "Wishiwashi", ability: 'schoolilng', item: 'blueorb', moves: ['uturn']}]);
```
Comment (Marty-D) [inline test/simulator/misc/mixandmega.js]: Schooling typo

**Comment by Marty-D** [comment]
Looks good, only other thing to consider is once those other files are set up for TypeScript, they'll need `// @ts-ignore` like the original `{id: 'transform'}` does or will fail `npm test` later. @Zarel should that just be done now or wait for the future?

**Comment by Zarel** [comment]
Wait for the future.

File: test/simulator/misc/megaevolution.js (hunk 1 above)
Comment (MacChaeger) [reply test/simulator/misc/megaevolution.js]: This was originally a loop that tested all nine but then Zarel mentioned offhand that he didn’t like those so I panicked and changed it to the first forme-change ability I could think of.

**Comment by MacChaeger** [comment]
I was actually thinking of just replacing that effect parameter with a Boolean because literally every single call to setAbility made in PS right now only uses that parameter to pass an id for these abilities, and absolutely none of them ever use that fourth parameter.

**Comment by MacChaeger** [comment]
When I say “every single” I mean including these changes. The current master only calls setAbility with three or more parameters one time and it’s to pass the `'transform'` id.

**Comment by Marty-D** [comment]
I thought `noForce` was used by Mummy? Did it stop using that?

**Comment by MacChaeger** [comment]
Ah, missed that.

**Comment by MacChaeger** [comment]
Wouldn't be too hard to replace that with a check to see if the attacker already has Mummy in Mummy's own code, though.

**Comment by Marty-D** [comment]
I agree that `setAbility` is kind of a mess in general though.

File: data/scripts.js [hunk 3]
```diff
@@ -897,7 +897,7 @@ exports.BattleScripts = {
 	},
 
 	runMegaEvo: function (pokemon) {
-		const isUltraBurst = !pokemon.canMegaEvo;
+		const effectType = !pokemon.canMegaEvo ? '-burst' : '-mega';
```
Comment (urkerab) [inline data/scripts.js]: Double negative might be confusing? I'd prefer `pokemon.canMegaEvo ? '-mega' : '-burst'`.

**Comment by Marty-D** [comment]
Don't forget the `setAbility` function from the Partners in Crime mod!

File: data/scripts.js [hunk 4]
```diff
@@ -897,7 +897,7 @@ exports.BattleScripts = {
 	},
 
 	runMegaEvo: function (pokemon) {
-		const isUltraBurst = !pokemon.canMegaEvo;
+		const effectType = pokemon.canMegaEvo ? '-mega' : '-burst';
```
Comment (Zarel) [inline data/scripts.js]: **I'D** prefer leaving it a boolean. Would have saved the recent huge bug and also would be more readable. Also `effectType` already means something else.

**Pull request merged**